task_manager/task_manager.hh

```diff
@@ -0,0 +1,323 @@
+
+        virtual is_abortable is_abortable() const noexcept {
+            return is_abortable::no;
+        }
+
+        void set_type(std::string type) noexcept {
+            _status.type = std::move(type);
+        }
+
+        void set_total_progress(double progress) {
+            _progress.total = progress;
+        }
+
+        void report_completion_progress(double progress) {
+            _progress.completed += progress;
+        }
+
+        void change_state(task_state state) noexcept {
+            _status.state = state;
+        }
+
+        virtual task_manager::task_ptr create_child(task_id id, task_manager::foreign_task_ptr&& parent = nullptr, unsigned sequence_number = 0, std::string keyspace = "", std::string table = "", std::string type = "", std::string entity = "") const = 0;
+
+        void start() {
+            _status.start_time = db_clock::now();
+            _status.state = task_manager::task_state::running;
+
+            (void)with_gate(_module->async_gate(), [f = done(), module = _module, id = _status.id] () mutable {
+                return std::move(f).finally([module, id] {
+                    return sleep_abortable(module->get_task_ttl(), module->abort_source());
+                }).then_wrapped([module, id] (auto f) {
+                    f.ignore_ready_future();
+                    module->unregister_task(id);
+                });
+            });
+        }
+
+        static future<std::vector<task_id>> run_on_shards(task_ptr task, module_ptr module) {
+            return module->_tm.container().map([parent = seastar::make_foreign(task), module] (task_manager&) mutable -> foreign_task_ptr {
+                task_ptr child = parent->create_child(task_id::create_random_id(), std::move(parent)); // TODO Deexie the rest of fields
+                module->register_task(child);
+                child->start();
+                return seastar::make_foreign<task_ptr>(child);
+            }).then([task] (foreign_task_vector tasks) {
+                task->_children = std::move(tasks);
+                return make_ready_future<std::vector<task_id>>(boost::copy_range<std::vector<task_id>>(task->_children
+                    | boost::adaptors::transformed([] (foreign_task_ptr& child) { return child->_status.id; })));
+            });
+        }
+
+        future<> done() const noexcept {
+            return _done.get_shared_future();
+        }
+
+        void unregister() {
+            _module->unregister_task(_status.id);
+        }
+
+        void finish() noexcept {
+            _status.end_time = db_clock::now();
+            _status.state = task_manager::task_state::done;
+            _done.set_value();
+        }
+
+        void finish_failed(std::exception_ptr ex, std::string error) noexcept {
+            _status.end_time = db_clock::now();
+            _status.state = task_manager::task_state::failed;
+            _status.error = std::move(error);
+            _done.set_exception(ex);
```

Comment:
I'm not sure we should do this.
When the task completes with an error we record the error status in _status.error (we can keep the exception_ptr in _status too if needed), but I'm not sure we want anyone waiting on `done()` to get an error in this case.
It might be beneficial to `set_value` here, in contrast to cases where, e.g. there was an higher level error to start the task and it didn't even run.